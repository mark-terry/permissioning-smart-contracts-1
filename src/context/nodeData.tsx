// Libs
import React, {
    createContext,
    useContext,
    useEffect,
    useState,
    useMemo
} from "react";
import { drizzleReactHooks } from "drizzle-react";

// Utils
import { paramsToIdentifier } from "../util/enodetools";

type Enode = {enodeHigh: string, enodeLow: string, identifier: string, ip: string, port: string};

type ContextType = {
    nodeWhitelist?: Enode[],
    setNodeWhitelist?: (enode: Enode[]) => void
}

const DataContext = createContext<ContextType>({});

/**
 * Provider for the data context that contains the whitelist
 * @param {Object} props Props given to the DataProvider
 * @return The provider with the following value:
 *  - nodeWhitelist: list of whiteliist enode from Node Rules contract
 *  - setNodeWhitelist: setter for the whitelist state
 */
export const NodeDataProvider: React.FC = (props: React.Props<{}>) => {
    const [nodeWhitelist, setNodeWhitelist] = useState<Enode[]>([]);

    const value = useMemo(() => ({ nodeWhitelist, setNodeWhitelist }), [
        nodeWhitelist,
        setNodeWhitelist
    ]);
    return <DataContext.Provider value={value} {...props} />;
};

export const useNodeData = () => {
    const context = useContext(DataContext);
    if (!context) {
        throw new Error("useData must be used within a DataProvider.");
    }

    const { nodeWhitelist, setNodeWhitelist } = context;
    const { drizzle, useCacheCall } = drizzleReactHooks.useDrizzle();    
    const nodeWhitelistSize: number = useCacheCall("NodeRules", "getSize");
    const { getByIndex: getNodeByIndex } = drizzle.contracts.NodeRules.methods;    
    const nodeIsReadOnly: boolean = useCacheCall("NodeRules", "isReadOnly");
    const { userAddress } = drizzleReactHooks.useDrizzleState((drizzleState: any) => ({
        userAddress: drizzleState.accounts[0]
    }));

    useEffect(() => {
        const promises = [];
        for (let index = 0; index < nodeWhitelistSize; index++) {
            promises.push(getNodeByIndex(index).call());
        }
        Promise.all(promises).then(responses => {
            const updatedNodeWhitelist = responses.map(
                ({ enodeHigh, enodeLow, ip, port }) => ({
                    enodeHigh,
                    enodeLow,
                    ip,
                    port,
                    identifier: paramsToIdentifier({
                        enodeHigh,
                        enodeLow,
                        ip,
                        port
                    })
                })
            );
            setNodeWhitelist!(updatedNodeWhitelist);
        });
    }, [nodeWhitelistSize, setNodeWhitelist, getNodeByIndex]);    

    const formattedNodeWhitelist = useMemo(() => {
        return nodeWhitelist
            ? nodeWhitelist
                  .map(enode => ({ ...enode, status: "active" }))
                  .reverse()
            : undefined;
    }, [nodeWhitelist]);    
    
    const dataReady = useMemo(
        () =>
            typeof nodeIsReadOnly === "boolean" &&
            Array.isArray(nodeWhitelist),
        [nodeIsReadOnly, nodeWhitelist]
    )

    return {
        userAddress,
        dataReady,
        whitelist: formattedNodeWhitelist,
        isReadOnly: nodeIsReadOnly
    }
}