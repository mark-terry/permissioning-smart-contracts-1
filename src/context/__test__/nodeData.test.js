// Libs
import React from "react";
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";
// Components
import { NodeDataProvider } from "../nodeData";

describe("<NodeDataProvider />", () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(
            <NodeDataProvider>
                <div className="test" />
            </NodeDataProvider>
        );
    });

    it("renders children when passed in", () => {
        expect(wrapper.contains(<div className="test" />)).toEqual(true);
    });

    it("matches snapshot", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
