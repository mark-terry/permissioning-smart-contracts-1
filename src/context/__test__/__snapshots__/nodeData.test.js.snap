// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NodeDataProvider /> matches snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "nodeWhitelist": Array [],
      "setNodeWhitelist": [Function],
    }
  }
>
  <div
    className="test"
  />
</ContextProvider>
`;
